### PR TITLE
Convert use_eval to use send/recv system

### LINKED ISSUE #1026: Async Communication Between JavaScript and Rust
## Specific Demand

Currently, it is difficult to integrate with some web or JavaScript APIs because use_eval only allows you to return a single JavaScript value.

For example, the [intersection observer API](https://developer.mozilla.org/en-US/docs/Web/API/Intersection_Observer_API) uses callbacks to notify a script when the intersection has changed, but there is no easy way to send values to rust from that callback.

Dioxus can implement some web APIs as first party. This somewhat mitigates this problem, but it cannot provide first party bindings for all the JavaScript libraries users may want to integrate.

## Implement Suggestion

Change use_eval to include an incoming and outgoing channel for both the JavaScript and Rust side. In JavaScript, a `dioxus` variable would be set before the script is run with methods to communicate with rust.

```rust
let eval = use_eval(cx);
let result = eval(r#"dioxus.send("hello from JS");
const value = await dioxus.recv();
console.log(value);"#);
println!("{}", result.recv());
result.send("hello");
```

### PR DESCRIPTION
**Pull request by DogeDark**
Converts the ``use_eval`` hook to use a send/recv system.

**Example:**
```rs
fn app(cx: Scope) -> Element {
    // Construct the evaluator
    let eval_provider = dioxus_html::prelude::use_eval(cx);
    let eval = eval_provider(r#"
        dioxus.send("Hi from JS!");
        let msg = await dioxus.recv();
        console.log(msg);
    "#);

    // Run the JS
    eval.run();

    // Send a message to JS
    eval.send("Hi from Rust!".into());

    // Get a receiver channel to receive messages from JS
    let receiver = eval.receiver();
    let future = use_future(cx, (), |_| async move {
        receiver.recv().await.unwrap()
    });
    
   // Act on received data
    match future.value() {
        Some(v) => {
            // Cleanup any eval artifacts
            eval.done();
            cx.render(rsx!(
                p { "{v}" }
            ))
        }
        _ => {
            cx.render(rsx!(
                p { "hello" }
            ))
        }
    }
}
```

**Todo:**
- [x] Web
- [x] Desktop
- [x] Liveview

Closes #1026

### CODE REVIEW

**Comment by DogeDark** [comment]
Will fix the CI issues tomorrow.

File: packages/html/src/eval.rs
Comment (DogeDark) [reply packages/html/src/eval.rs:45]: Done!

File: examples/eval.rs
Comment (Demonthos) [inline examples/eval.rs:22]: Running this example on my M1 mac, it panics unwrapping at this line

File: packages/web/src/eval.js
Comment (Demonthos) [inline packages/web/src/eval.js:10]: Instead of polling for new data could we first check if there is any currently resolved data. If there is we can resolve the promise immediately, otherwise we could set a pendingPromise field on the object to the resolve callback. That callback could then be called in the rustSend function.

File: examples/eval.rs
Comment (DogeDark) [reply examples/eval.rs:22]: Does it say why?

Comment (DogeDark) [reply examples/eval.rs:22]: Nevermind, I'm getting it too.

File: packages/web/src/eval.js
Comment (DogeDark) [reply packages/web/src/eval.js]: Much better solution! I knew a timeout didn't feel right.

File: examples/eval.rs
Comment (DogeDark) [reply examples/eval.rs:22]: Should be fixed now.

File: packages/html/src/eval.rs
Comment (Demonthos) [inline packages/html/src/eval.rs:33]: Returning &mut UseEval here where eval is not clone makes it [difficult to move around UseEval](https://dioxuslabs.com/docs/nightly/guide/en/interactivity/custom_hooks.html#hook-anti-patterns). I think you should still be able to start an eval within a future with this API, but you would need to make the javascript await for a start message after you run it which could be unintuitive

I also wonder if we could make the API a bit more similar to the previous API to make upgrading less breaking. What are your thoughts on this API?
- use_eval returns a Rc<dyn EvalProvider>
- You can call `.start` on the eval provider to start running the script
- The Eval returned from `.start` can either be turned into a future with [IntoFuture](https://doc.rust-lang.org/nightly/core/future/trait.IntoFuture.html) (which consumes it) like the old API or you can use the send/receive function like you currently implemented

Comment (DogeDark) [reply packages/html/src/eval.rs:33]: I agree with the Rc but instead of a start function, how about implementing ``IntoFuture`` for the  ``UseEval`` struct? It would automatically run and await the first data returned through the channel. 

Additionally, any ``return data;`` JS could be replaced with ``dioxus.send(data);``? - it would add some string processing overhead during runtime though. If not that then we could add it as a step in the migration process - tell people to replace ``return`` with ``dioxus.send``?

```rs
// I could see this being useful outside of legacy code
let eval = use_eval(cx, "dioxus.send('hi from js');");
if let Ok(msg) = eval.await {
    // act on message
}
```
compared to the old:
```rs
let eval = use_eval(cx);
if let Ok(msg) = eval("return 'hi from js';").await {
    // act on message
}
```

Comment (Demonthos) [reply packages/html/src/eval.rs:33]: > I agree with the Rc but instead of a start function, how about implementing `IntoFuture` for the `UseEval` struct? It would automatically run and await the first data returned through the channel.

This sounds good!

> Additionally, any `return data;` JS could be replaced with `dioxus.send(data);`? - it would add some string processing overhead during runtime though. (**On second thoughts, replacing the return statement could have false positives**)

Nested function make this difficult. I think it would be better to try to not modify the javascript script submitted (except the wrapping we already do).

> If not that then we could add it as a step in the migration process - tell people to replace `return` with `dioxus.send`?
> 
> ```rust
> // I could see this being useful outside of legacy code
> let eval = use_eval(cx, "dioxus.send('hi from js');");
> if let Ok(msg) = eval.await {
>     // act on message
> }
> ```
> 
> compared to the old:
> 
> ```rust
> let eval = use_eval(cx);
> if let Ok(msg) = eval("return 'hi from js';").await {
>     // act on message
> }
> ```

I could see the old system being useful when you have a chain of several different scripts like:

```rust
let eval = use_eval(cx);
use_future!(cx, |()| {
    to_owned![eval];
    async move {
        eval("return 'script 1'").await;
        // some rust processing code
        eval("return 'script 2'").await;
    }
});
```
Where script 1 and script 2 are from different libraries.

In the new system, that would need to look like this:

```rust
let eval1 = use_eval(cx, "script 1");
let eval2 = use_eval(cx, "script 2");
use_future!(cx, |()| {
    to_owned![eval1, eval2];
    async move {
        eval1.await;
        // some rust processing code
        eval2.await;
    }
});
```

If we force the code to be defined with the hook it could also hurt colocation:

```rust
let script = use_eval(cx, "script 1");

// 100 more lines of code setting up data

render!{
    button {
        onclick: move |_| async move {
            script.await
        }
    }
}
```

Looking at the onclick handler, it could be less clear what code script.await is running compared to:

```rust
let eval = use_eval(cx);

// 100 more lines of code setting up data

render!{
    button {
        onclick: move |_| async move {
            eval("script 1").await
        }
    }
}
```

Comment (DogeDark) [reply packages/html/src/eval.rs:33]: > Looking at the onclick handler, it could be less clear what code script.await is running compared to: 
> ```rs
> let eval = use_eval(cx);
> 
> // 100 more lines of code setting up data
> 
> render!{
>     button {
>         onclick: move |_| async move {
>             eval("script 1").await
>         }
>     }
> }
> ```

I believe this shouldn't be as big of an issue with good variable naming. Developers could also move the use_eval call below the 100 lines.

> In the new system, that would need to look like this:
> ```rs
> let eval1 = use_eval(cx, "script 1");
> let eval2 = use_eval(cx, "script 2");
> use_future!(cx, |()| {
>     to_owned![eval1, eval2];
>     async move {
>         eval1.await;
>         // some rust processing code
>         eval2.await;
>     }
> });
> ```
I agree, chaining eval is useful. However, what if something like this was implemented which could effectively do the same:
```rs
let eval = use_eval(cx, 
    r#"
    script 1
    // wait for signal to continue
    await dioxus.signal(); // reads as "await dioxus signal"
    script 2
    "#);

use_future!(cx, |()| {
    to_owned![eval];
    async move {
        eval.run();
        // some rust processing code
        eval.signal(); // Signal the script to continue
    }
});
```
The dev could still pass data to the script when needed using the regular ``send()`` and ``recv()``.

I generally want to avoid using any closures or anonymous functions. In my opinion, they add complexity and make control flow a little confusing (dioxus-std was a pain).

Comment (DogeDark) [reply packages/html/src/eval.rs]: I did the stuff @Demonthos
This works
```rs
let mut eval = use_eval(cx);

// 100 more lines of code setting up data

render!{
    button {
        onclick: move |_| async move {
            eval("script 1").await
        }
    }
}
```

File: packages/desktop/src/query.rs
Comment (Demonthos) [inline packages/desktop/src/query.rs:106]: This ends up scaling poorly with large numbers of requests (explained more in https://github.com/DioxusLabs/dioxus/pull/1115). This shouldn't be any worse than the git version of use_eval. At this point I would like to get this PR merged, so that we can unblock the liveview router. Just something to think about for the future

File: playwrite-tests/web/src/main.rs
Comment (Demonthos) [inline playwrite-tests/web/src/main.rs:55]: Could we just return an async block here or use cx.spawn instead of pulling in another library?

**Review by Demonthos — CHANGES_REQUESTED** [review]
API looks great! A few small things. I think we should wait on performance improvements so that we can get this into 0.4, but if you could remove `wasm_bindgen_futures::spawn_local` that would be great

File: playwrite-tests/web/src/main.rs
Comment (DogeDark) [reply playwrite-tests/web/src/main.rs]: Turns out we don't even need async.

File: examples/eval.rs
Comment (Demonthos) [reply examples/eval.rs:22]: This seems like it became an issue again. If the eval gets dropped it will panic

This fixes the issue, but if it is possible to make the original example work, that would probably be better:

```rust
use dioxus::prelude::*;

fn main() {
    dioxus_desktop::launch(app);
}

fn app(cx: Scope) -> Element {
    let eval_provider = use_eval(cx);

    let future = use_future(cx, (), |_| {
        let mut eval = eval_provider(
            r#"
            dioxus.send("Hi from JS!");
            let msg = await dioxus.recv();
            console.log(msg);
        "#,
        );
        async move {
            eval.run().unwrap();
            eval.send("Hi from Rust!".into()).unwrap();

            let receiver = eval.receiver();
            receiver.recv().await.unwrap()
        }
    });

    match future.value() {
        Some(v) => cx.render(rsx!(
            p { "{v}" }
        )),
        _ => cx.render(rsx!(
            p { "hello" }
        )),
    }
}
```

Comment (DogeDark) [reply examples/eval.rs:22]: Realistically the error should be handled, but we could create a wrapper type around the channel that has the same methods but instead of errors it returns Option<Value>

**Comment by dr-oofenshmirtz** [comment]
Sorry to hijack. I'm using this PR and I am trying to get multiple messages from the channel. Having replicated your example successfully. I set out to tweak it for my use case. 

```rust
     let mut eval = eval_provider(
         r#"
           window.addEventListener("message", dioxus.send)
    "#,);
```

Which should in theory attach an event listener to the window. Listening for a postMessage from either a service worker or an iframe. (technically the same context works, but that's not the value add). The issue I'm having is the receiver is falling out of scope. In [the example you create a receiver](https://github.com/DioxusLabs/dioxus/pull/1080/files#diff-7c4e660449b789afbf10c871b80773475ec4eb758cfd01e7c0f50b6ce3f5bbabR20) and you use it as a future. If I grok the code this is specifically an [UnboundedReciever](https://docs.rs/futures-channel/latest/futures_channel/mpsc/struct.UnboundedReceiver.html#). Which leads me to believe that if I stuck it in a variable or a use_ref or some other form of maintaining a reference to the channel, I could read from it indefinitely. Which should work perfectly with window.addEventListener. 

Am I understanding the potential correctly? How would you recommend I maintain a reference to the channel to get multiple values out of it? And bonus points, if I make an event listener in the JS context in this manner, how could I clean it up?

**Comment by DogeDark** [comment]
Yes, I believe that's possible.

> I could read from it indefinitely.

It sounds like you want [cx.spawn](https://docs.rs/dioxus/latest/dioxus/prelude/struct.ScopeState.html#method.spawn),  [cx.spawn_forever](https://docs.rs/dioxus/latest/dioxus/prelude/struct.ScopeState.html#method.spawn_forever) or maybe even [use_coroutine](https://docs.rs/dioxus-hooks/latest/dioxus_hooks/fn.use_coroutine.html). You would move the eval into the async block (solving the drop-ing issue) and then call ``eval.run()`` on it (so it only runs once and not every time the component re-renders).

In terms of cleaning up, you could have the script wait for a message before running additional code:
```rs
let mut eval = eval_provider(r#"
    window.addEventListener("message", dioxus.send);
    
    // Wait for any message from Rust
    await dioxus.recv();

    // Cleanup code
    window.removeEventListener("message", dioxus.send);
    "#);

// When you want to cleanup the event listener:
eval.send(serde_json::Value::Null);
```

**Comment by dr-oofenshmirtz** [comment]
> @dr-oofenshmirtz Yes, I believe that's possible.
> 
> > I could read from it indefinitely.
> 
> It sounds like you want [cx.spawn](https://docs.rs/dioxus/latest/dioxus/prelude/struct.ScopeState.html#method.spawn), [cx.spawn_forever](https://docs.rs/dioxus/latest/dioxus/prelude/struct.ScopeState.html#method.spawn_forever) or maybe even [use_coroutine](https://docs.rs/dioxus-hooks/latest/dioxus_hooks/fn.use_coroutine.html). You would move the eval into the async block (solving the drop-ing issue) and then call `eval.run()` on it (so it only runs once and not every time the component re-renders).
> 
> In terms of cleaning up, you could have the script wait for a message before running additional code:
> 
> ```rust
> let mut eval = eval_provider(r#"
>     window.addEventListener("message", dioxus.send);
>     
>     // Wait for any message from Rust
>     await dioxus.recv();
> 
>     // Cleanup code
>     window.removeEventListener("message", dioxus.send);
>     "#);
> 
> // When you want to cleanup the event listener:
> eval.send(serde_json::Value::Null);
> ```

Clever use of await!

**Review by Demonthos — APPROVED** [review]
Looks great! (waiting for a second review before merging)

**Comment by Demonthos** [comment]
This now works with Liveview, and it implements the more efficient non-broadcast method of channels

**Pull request merged**
